internal/provider/resource_application.go

```diff
@@ -578,7 +578,15 @@ func (r *applicationResource) Create(ctx context.Context, req resource.CreateReq
 	r.trace(fmt.Sprintf("read application resource %q", createResp.AppName))
 
 	// Save plan into Terraform state
-	plan.Constraints = types.StringValue(readResp.Constraints.String())
+
+	// Constraints do not apply to subordinate applications. However, whilst juju
+	// may not distinguish between null constraints and empty-string constraints,
+	// terraform does not consider null to be a computed value.
+	if readResp.Principal {
+		plan.Constraints = types.StringValue(readResp.Constraints.String())
+	} else {
+		plan.Constraints = types.StringValue("")
+	}
```

Comment:
If this is a subordinate, readResp.Constraints will be the empty string anyway

Perhaps we should drop this if statement drop this business logic entirely out of the clients?

Reply:
literally just
```
plan.Constraints = types.StringValue(readResp.Constraints.String())
```